temporal-serviceclient/src/main/java/io/temporal/serviceclient/SystemInfoInterceptor.java

```diff
@@ -1,104 +0,0 @@
-
-  public SystemInfoInterceptor(
-      CompletableFuture<GetSystemInfoResponse.Capabilities> serverCapabilitiesFuture) {
-    this.serverCapabilitiesFuture = serverCapabilitiesFuture;
-  }
-
-  @Override
-  public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
-      MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
-    return new ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(
-        next.newCall(method, callOptions)) {
-
-      @Override
-      public void start(Listener<RespT> responseListener, Metadata headers) {
-        if (!serverCapabilitiesFuture.isDone()) {
-          if (method == WorkflowServiceGrpc.getGetSystemInfoMethod()) {
-            // It's already a getSystemsInfo call, so we just listen on it and populate the
-            // capabilities
-            responseListener =
-                new ForwardingClientCallListener.SimpleForwardingClientCallListener<RespT>(
-                    responseListener) {
-                  @Override
-                  public void onMessage(RespT message) {
-                    if (message instanceof GetSystemInfoResponse) {
-                      GetSystemInfoResponse response = (GetSystemInfoResponse) message;
-                      serverCapabilitiesFuture.complete(response.getCapabilities());
-                    }
-                    super.onMessage(message);
-                  }
-
-                  @Override
-                  public void onClose(Status status, Metadata trailers) {
-                    if (Status.UNIMPLEMENTED.getCode().equals(status.getCode())) {
-                      serverCapabilitiesFuture.complete(
-                          GetSystemInfoResponse.Capabilities.getDefaultInstance());
-                    }
-                    super.onClose(status, trailers);
-                  }
-                };
-          } else {
-            // Need to reach system capabilities, so make a getSystemInfo call in a blocking manner.
-            // We don't try to squash into one and optimize the several getSystemInfo calls that may
-            // be initiated by several client calls here. Doing so it will require tricky
-            // implementation to ensure proper deadlines that may be different between calls.
-            // If a server is able to take the load of the requests, it should be able to serve some
-            // additional lightweight static getSystemInfo calls that are serialized with the actual
-            // calls.
-            serverCapabilitiesFuture.complete(
-                getServerCapabilitiesOrThrow(next, callOptions.getDeadline()));
-          }
-        }
-
-        super.start(responseListener, headers);
-      }
-    };
-  }
-
-  public static GetSystemInfoResponse.Capabilities getServerCapabilitiesOrThrow(
-      Channel channel, @Nullable Deadline deadline) {
-    try {
-      return WorkflowServiceGrpc.newBlockingStub(channel)
```

Comment:
I think we just need to add a retried with a deadline here no?

Reply:
???

I thought we'd discussed the fact that no other language SDK has an interceptor for this purpose.

In particular, after studying it I don't believe this interceptor ever does anything, ever, since all `CompletableFuture.complete` calls after the first will be ignored.

Reply:
On second look, are you speaking specifically of this method?  I moved it to ChannelManager and renamed it to `getServerCapabilitiesRaw`, then wrapped it with a retry.  I already did what you're asking for?

Reply:
My suggestion was to keep to simple and just retry this call and don't modify anything else.

Reply:
>I thought we'd discussed the fact that no other language SDK has an interceptor for this purpose.

`getServerCapabilities` is supposed to be called before any RPC, an interceptor is a way to accomplish that and I do believe it does that.

Reply:
> > I thought we'd discussed the fact that no other language SDK has an interceptor for this purpose.
> 
> `getServerCapabilities` is supposed to be called before any RPC, an interceptor is a way to accomplish that and I do believe it does that.

Um, what, no?  It hooks into existing calls to `WorkflowService.getSystemInfo` to capture the outputs.  It does not provoke additional calls to `WorkflowService.getSystemInfo`.  Have you ever read the code?

Reply:
But there is an if condition that makes a call no?
```
if (method == WorkflowServiceGrpc.getGetSystemInfoMethod()) {
...
} else {
            // Need to reach system capabilities, so make a getSystemInfo call in a blocking manner.
            // We don't try to squash into one and optimize the several getSystemInfo calls that may
            // be initiated by several client calls here. Doing so it will require tricky
            // implementation to ensure proper deadlines that may be different between calls.
            // If a server is able to take the load of the requests, it should be able to serve some
            // additional lightweight static getSystemInfo calls that are serialized with the actual
            // calls.
            serverCapabilitiesFuture.complete(
                getServerCapabilitiesOrThrow(next, callOptions.getDeadline()));
          }
}
```

If the call is not `getGetSystemInfoMethod` won't we call `getServerCapabilitiesOrThrow()` which does make an an additional call to `WorkflowService.getSystemInfo`?

Reply:
That code path is only executed if `!serverCapabilitiesFuture.isDone()`.  It is done when `ChannelManager.connect` calls `new GrpcRetryer(getServerCapabilities()).retryWithResult`, which calls `Supplier.get` before entering the retry loop, which completes the future in the lambda body of `ChannelManager.getServerCapabilities`.  I guarantee to you that this code path will never execute if you put a logging statement in it.